### PR TITLE
noderesourcetopology: overhaul Topology Manager configuration management

### PR DESCRIPTION
**Pull request by ffromani**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
The NodeResourceTopology API is deprecating the `TopologyPolicies` field, which is the way the agents (NFD) exported the relevant Topology Manager options (see: https://github.com/k8stopologyawareschedwg/noderesourcetopology-api/issues/24 https://github.com/k8stopologyawareschedwg/noderesourcetopology-api/pull/25).
The new recommended option is to expose the configuration using Attributes.
Add relevant code to generalize the configuration management while keep backward compatibility with the current generation of updaters

#### Which issue(s) this PR fixes:
Fixes N/A

#### Special notes for your reviewer:
N/A

#### Does this PR introduce a user-facing change?
```release-note
the NodeTopologyMatch plugins now depend on v1alpha2 version of the noderesourcetopology API.
```

### CODE REVIEW

**Comment by ffromani** [comment]
/hold cancel
@swatisehgal: do we want to wait for NRT v1alpha2 and put this PR on hold up until that or do we want to go with this fake-zone approach? (I have a _small_ preference towards the fake-zone approach because it becomes kinda like a namespace, even if arguably we're [almost] abusing the zone concept)
@PiotrProkop this is the general direction we are pursuing after a conversation with NFD folks. Could be either top-level attributes (we will need v1alpha2 for that, see the last updates on https://github.com/k8stopologyawareschedwg/noderesourcetopology-api/pull/25) or this fake-zone approach, but we agreed attributes are better fit than labels/annotations.

**Comment by PiotrProkop** [comment]
@ffromani thanks for the info. I'll review this tomorrow.

**Comment by swatisehgal** [comment]
> /hold cancel @swatisehgal: do we want to wait for NRT v1alpha2 and put this PR on hold up until that or do we want to go with this fake-zone approach? (I have a _small_ preference towards the fake-zone approach because it becomes kinda like a namespace, even if arguably we're [almost] abusing the zone concept) 

I would prefer to get `v1alpha2` merged first primarily for the reasons that you have pointed out: it seems like an abuse of the zone concepts. In addition to that, it seems like a workaround when we have already agreed on a clear way forward. If we were to go ahead with this change, we would have to cleanup once we would transition to v1alpha2. I would rather us do it the right way in one shot.

We can easily unblock the transition to v1alpha2. I noticed that the PR is WIP and that is why never reviewed. I will take a look and then we can update this PR to consume v1alpha2 version of the API. 

Sounds reasonable?

**Comment by ffromani** [comment]
/hold
let's wait for `v1alpha2` then!

**Comment by Huang-Wei** [comment]
@swatisehgal @ffromani Just a heads-up, once you proceed to bump the version of NRT, please include a release-note to explain its potential behavior change in brief. 🙏

**Comment by ffromani** [comment]
> @swatisehgal @ffromani Just a heads-up, once you proceed to bump the version of NRT, please include a release-note to explain its potential behavior change in brief. pray

ACK, I'll need to revisit the `release-note` field

**Comment by PiotrProkop** [comment]
> > /hold cancel @swatisehgal: do we want to wait for NRT v1alpha2 and put this PR on hold up until that or do we want to go with this fake-zone approach? (I have a _small_ preference towards the fake-zone approach because it becomes kinda like a namespace, even if arguably we're [almost] abusing the zone concept)
> 
> I would prefer to get `v1alpha2` merged first primarily for the reasons that you have pointed out: it seems like an abuse of the zone concepts. In addition to that, it seems like a workaround when we have already agreed on a clear way forward. If we were to go ahead with this change, we would have to cleanup once we transition to v1alpha2. I would rather us do it the right way in one shot.
> 
> We can easily unblock the transition to v1alpha2. I noticed that the PR is WIP and that is why never reviewed. I will take a look and then we can update this PR to consume v1alpha2 version of the API.
> 
> Sounds reasonable?

I agree with @swatisehgal that we should wait for `v1alpha2` not to abuse `fake` zone concept(such hacks have tendencies to linger for a long time).

File: pkg/noderesourcetopology/config.go
Comment (PiotrProkop) [inline pkg/noderesourcetopology/config.go:91]: Can we add `TODO` to add `TopologyManagerPolicies` when `scheduler-plugins` supports k8s 1.26. Basically when this [PR](https://github.com/kubernetes-sigs/scheduler-plugins/pull/489) is merged

Comment (PiotrProkop) [reply pkg/noderesourcetopology/config.go:91]: Yeah, my bad :) comment updated

Comment (ffromani) [reply pkg/noderesourcetopology/config.go:91]: sure thing, it was planned already (https://github.com/kubernetes-sigs/scheduler-plugins/pull/488/commits/b131bdada31d1be1547beab439a4b1b5476cad17#diff-3532db6795688cc2309165cf54c8ad8d256cb477ca3f9a5b0f778f85aabf1b49R132) but forgot to add TODOs here

File: manifests/noderesourcetopology/crd.yaml
Comment (PiotrProkop) [inline manifests/noderesourcetopology/crd.yaml:6]: I think this annotation has to stay to be able to create CRD in `k8s.io` domain

File: pkg/noderesourcetopology/config.go
Comment (PiotrProkop) [inline pkg/noderesourcetopology/config.go:63]: I have a question, it is valid not specify both `scope` and `policy` in Kubelet config, and their will be defaulted to `container` for scope and `none` for policy. Do we assume that NFD(or whichever component will advertise those config option) will make sure to make explicit conversion from "" to default value or we should just assume here that those values weren't set and we can just change "" to  default values here.

File: manifests/noderesourcetopology/crd.yaml
Comment (ffromani) [reply manifests/noderesourcetopology/crd.yaml:6]: I agree. But it should be added automatically (not manually) if possible, hence: https://github.com/k8stopologyawareschedwg/noderesourcetopology-api/pull/29

File: pkg/noderesourcetopology/config.go
Comment (ffromani) [reply pkg/noderesourcetopology/config.go:63]: that's a good point. I don't have an answer yet for NFD, and I'm revisiting this function and how we use it.

File: manifests/noderesourcetopology/crd.yaml
Comment (PiotrProkop) [reply manifests/noderesourcetopology/crd.yaml:6]: Ok, I can see you got this covered 😄

Comment (ffromani) [reply manifests/noderesourcetopology/crd.yaml:6]: hopefully! :)

File: pkg/noderesourcetopology/config.go
Comment (ffromani) [reply pkg/noderesourcetopology/config.go:63]: These details should be discussed and documented, because they are a key part of the producer-consumer contract. I started a change (in this PR) to add this documentation in `pkg/noderesourcetopology/README.md`, let's continue here. If we find we need more space and discussion, we can move later to another forum (e.g. gh issue, google doc...)

Comment (PiotrProkop) [reply pkg/noderesourcetopology/config.go:63]: Yeah, I just feel producer should not do a lot of preprocessing so we can just emit what's configured in kubelet and not bother about complicated logic. IMHO setting defaults for those values should happen in consumer(in this case NodeResourceTopologyMatch plugin).

Comment (ffromani) [reply pkg/noderesourcetopology/config.go:63]: > Yeah, I just feel producer should not do a lot of preprocessing so we can just emit what's configured in kubelet and not bother about complicated logic. IMHO setting defaults for those values should happen in consumer(in this case NodeResourceTopologyMatch plugin).

yes, I'm converging towards this point. What I think now (subject to change/further thinking):

- producers (NFD topology updater) should try their hardest to provide the values, but it the kubelet doesn't provide sufficient introspection mechanism, they should NOT guess or worse hardcode values. I think we're already in agreement here.
- consumer (scheduler plugin) CAN have defaults, because out of necessity we duplicate TM logic and constants anyway. But I'm still reluctant to hardcode defaults here, let's see.

Comment (PiotrProkop) [reply pkg/noderesourcetopology/config.go:63]: >     * consumer (scheduler plugin) CAN have defaults, because out of necessity we duplicate TM logic and constants anyway. But I'm still reluctant to hardcode defaults here, let's see.

No biggie for us when it comes with empty `policy` NodeResourceTopologyMatch  is noop for `none` policy. But we should probably assume that if `scope` is empty it means it is `container` scope(I bet many Kubernetes configurations uses TopologyManager but do not set `scope` and just rely on default here)

Comment (ffromani) [reply pkg/noderesourcetopology/config.go:63]: I'm not fully sold yet, but hardcoding default policy=none and default scope=container and using them as fallback if nothing is given from the producer (NFD) is something we can surely try out and evaluate.

Comment (ffromani) [reply pkg/noderesourcetopology/config.go:65]: Added in the last upload as promised

**Comment by ffromani** [comment]
still WIP: even though we expect no major changes left for NRT API v1alpha2, let's keep this PR on hold until we have a new tag for the NRT API. I'll finish the PR once we have that.

**Comment by ffromani** [comment]
the PR is now reviewable again

**Comment by ffromani** [comment]
/hold cancel
we have v1apha2 API: https://github.com/k8stopologyawareschedwg/noderesourcetopology-api/releases/tag/v0.1.0

File: pkg/noderesourcetopology/score.go
Comment (PiotrProkop) [inline pkg/noderesourcetopology/score.go:164]: I think you can just call `getScoringStrategyFunction` func here and pass it down. By doing it you can remove `scoreStrategyFunc` field from `TopologyMatch` struct, but it would mean that either this function will return additional `error` or we should add some validation in `TopologyMatch` constructor, WDYT?

**Comment by PiotrProkop** [comment]
And can you update compatibility Matrix? 😄

**Comment by ffromani** [comment]
> And can you update compatibility Matrix? smile

absolutely, thanks for reminding me!

File: pkg/noderesourcetopology/config.go
Comment (ffromani) [reply pkg/noderesourcetopology/config.go:101]: thanks, fixed!

File: pkg/noderesourcetopology/score.go
Comment (ffromani) [reply pkg/noderesourcetopology/score.go:164]: ok, unfortunately I have to take it back, this approach looks nicer and carrying `scoreStrategyFunc` is unfortunate, but it allows us to do the validation in the` TopologyMatch` constructor. Erroring out later in `Score` is too late and can only lead to silent failures - no way to abort, we can only set the score. Let me at least document this rationale with a code comment.

Comment (PiotrProkop) [reply pkg/noderesourcetopology/score.go:164]: Another thing we can try is to implement suggestion from this comment https://github.com/kubernetes-sigs/scheduler-plugins/pull/474#discussion_r1086235462, I think this can vastly simplify things in the future :) or I can just make an issue to add args validation for `TopologyMatch` plugin.

**Comment by ffromani** [comment]
/hold
let us wait until NFD master branch is updated to work with v1alpha2

**Review by PiotrProkop — COMMENTED** [review]
/lgtm 
@swatisehgal for proper LGTM 🚀

File: pkg/noderesourcetopology/score.go
Comment (PiotrProkop) [reply pkg/noderesourcetopology/score.go:164]: 👍

File: pkg/noderesourcetopology/README.md
Comment (PiotrProkop) [inline pkg/noderesourcetopology/README.md:170]: and I found the typo
```suggestion
  - example: `topologyManagerPolicy` becomes `topologyManagerPolicy`
```

Comment (ffromani) [reply pkg/noderesourcetopology/README.md]: good catch, thanks! fixed

**Comment by ffromani** [comment]
depends on https://github.com/kubernetes-sigs/node-feature-discovery/pull/1054 (just completing the cross-link reference)

**Comment by ffromani** [comment]
/hold cancel
both NFD and RTE have been updated

**Comment by ffromani** [comment]
@PiotrProkop @swatisehgal 
did minor cleanups/enhancements since last LGTM, could you PTAL again when you have some time?

**Comment by PiotrProkop** [comment]
LGTM

File: pkg/noderesourcetopology/config.go
Comment (swatisehgal) [inline pkg/noderesourcetopology/config.go:75]: IIUC, we need both these ways of obtaining topology policies as the exporter still might be using the `topologyPolicies` field for exposing information and we want to continue supporting that as  a precautionary measure. We are intentionally taking the route of soft deprecation and not fully deprecating the use of `topologyPolicies` in the v1alpha2 version of NRT API . Let's emphasize this and the fact that in the next iteration (when we move to v1beta1 version) we would be deprecating the use of `topologyPolicies` field.

File: test/integration/noderesourcetopology_cache_test.go
Comment (swatisehgal) [inline test/integration/noderesourcetopology_cache_test.go:545]: I am really happy about the use of attributes for capturing podfingerprint. It is much nicer than the use of annotations we did previously!

File: pkg/noderesourcetopology/README.md
Comment (swatisehgal) [inline pkg/noderesourcetopology/README.md:163]: Just a small rephrase: 
```suggestion
In addition to logically partitioning a cluster like explained above, Topology-aware scheduler needs to know key node-specific configuration settings like Topology manager policy and scope.
```

File: pkg/noderesourcetopology/README.md
Comment (swatisehgal) [inline pkg/noderesourcetopology/README.md:179]: NIT:
``` suggestion
    in NRT objects, and this requirement could be lifted in the future (an upgrade path will be provided).
```

**Review by swatisehgal — COMMENTED** [review]
Thanks for your work on this PR!

I left some minor comments and pointed out some NITs but overall the PR looks good to me.

File: pkg/noderesourcetopology/objects.go
Comment (Tal-or) [inline pkg/noderesourcetopology/objects.go:60]: We can use functional options here instead https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis 
But probably in separate PR

File: pkg/noderesourcetopology/objects.go
Comment (Tal-or) [inline pkg/noderesourcetopology/objects.go:74]: dito

**Review by Tal-or — COMMENTED** [review]
Thank you for the PR.
LGTM

File: pkg/noderesourcetopology/config.go
Comment (ffromani) [reply pkg/noderesourcetopology/config.go:42]: good idea, done!

File: pkg/noderesourcetopology/config.go
Comment (ffromani) [reply pkg/noderesourcetopology/config.go:75]: Right. Added a code comment *AND* a warning

File: pkg/noderesourcetopology/README.md
Comment (ffromani) [reply pkg/noderesourcetopology/README.md:163]: thanks, fixed

File: pkg/noderesourcetopology/README.md
Comment (ffromani) [reply pkg/noderesourcetopology/README.md:179]: thanks, fixed

File: pkg/noderesourcetopology/objects.go
Comment (ffromani) [reply pkg/noderesourcetopology/objects.go:60]: I tend to agree, but let's indeed defer to a separate PR.

**Review by swatisehgal — APPROVED** [review]
/lgtm

/assign @Huang-Wei 
for approval

**Comment by Huang-Wei** [comment]
/approve

Thanks @ffromani and @swatisehgal @Tal-or @PiotrProkop for the thorough review!

**Pull request merged**
